docs: add monitoring & observability guide (#198)
ClickHouse/clickhouse-operator: docs/guides/monitoring.mdx

---
position: 3
slug: /clickhouse-operator/guides/monitoring
title: 'Monitoring'
keywords: ['kubernetes', 'prometheus', 'monitoring', 'metrics']
description: 'How to scrape, secure, and use the operator metrics and health endpoints.'
doc_type: 'guide'
---

# Monitoring the ClickHouse Operator

The operator exposes Prometheus-compatible metrics and Kubernetes health probes so that you can observe its reconciliation activity, detect stalled controllers, and alert on failures.

This guide covers what the operator exposes, how to scrape it, and which queries are useful day to day.

<Note>

> This guide is about the **operator process itself** (the controller manager). For ClickHouse server metrics (queries, parts, replication lag), enable the [Prometheus endpoint in ClickHouse](/operations/server-configuration-parameters/settings#prometheus) via `extraConfig` and scrape it separately.

Reviewer: Prometheus port in the ClickHouse is enabled by default
```suggestion
This guide is about the **operator process itself** (the controller manager). For ClickHouse server metrics (queries, parts, replication lag), use the [Prometheus endpoint in ClickHouse](/operations/server-configuration-parameters/settings#prometheus) to scrape it separately.
```

```yaml
# values.yaml — minimal override
prometheus:
  enable: true
```

If you do not use cert-manager, additionally set `certManager.enable: false` and the ServiceMonitor will scrape with `insecureSkipVerify: true`, relying on bearer-token authentication only.

The full set of metrics-related defaults is:

```yaml
metrics:
  enable: true
  port: 8080
  secure: true            # HTTPS with authn/authz enforced on every scrape

certManager:
  enable: true            # Issues the metrics server certificate

prometheus:
  enable: false           # Set to true to render the ServiceMonitor
  scraping_annotations: false   # Alternative: prometheus.io/scrape pod annotations
```

Apply:

```bash
helm upgrade --install clickhouse-operator \
  oci://ghcr.io/clickhouse/clickhouse-operator-helm \
  -n clickhouse-operator-system --create-namespace \
  -f values.yaml
```

After install the chart creates:

- `Service/<release>-controller-manager-metrics-service` — exposes port `8080` (HTTPS when `metrics.secure: true`).
- `ServiceMonitor/<release>-controller-manager-metrics-monitor` — when `prometheus.enable: true`.

> - `ClusterRole/clickhouse-operator-metrics-reader` — non-resource URL `/metrics` with `get` verb.

Reviewer: ```suggestion
- `Service/<resource-prefix>metrics-service` — exposes port `8080` (HTTPS when `metrics.secure: true`).
- `ServiceMonitor/<resource-prefix>-controller-manager-metrics-monitor` — when `prometheus.enable: true`.
- `ClusterRole/<resource-prefix>-metrics-reader` — non-resource URL `/metrics` with `get` verb.
```
